### ISSUE TITLE
Question about the scripts

### ISSUE THREAD

**Issue by ge35tay** [matthewoots/upg-simulator#1]
Hi, I am new to Unity and want to ask a thing. Can u please explain which script is to attach to which Gameobject? This is really an interesting project reference for me and can u have a more detail comment in the readme?  Thanks advance!

**Comment by matthewoots** [comment]
Hi @ge35tay, thanks for checking out this repo, and glad that you are starting on Unity3d. Maybe you can let me know which **gameobjects** and **scripts** or what **functions** in the package that you would want to know, then I can address it here.😊 Any questions are welcomed.

**Comment by ge35tay** [comment]
Thank you very much! The first question is that when I import this project to the unity, it says there is a compilation error in it, 
![微信图片_20220313093809](https://user-images.githubusercontent.com/65305789/158051777-c8f108b5-554b-4bfa-9812-53ad7ef56466.png)

and I tried to play it in the unity and the following error in the console
![微信图片_20220313094041](https://user-images.githubusercontent.com/65305789/158051856-63e32cf9-7b5b-4745-98dc-b4f15cf86717.png)
Should I set up the ROS-TCP-Connecter at first? Which Ros message path should I choose in the unity robotics?

The second question is, I want to try to control DjiFPV by ROS. But I am not sure which script should add as a component in the Gameobjects of Dji model,  so that it can publish its position?

**Comment by matthewoots** [comment]
Sorry about this, I think I didn't add some instructions, but yes please follow `ROS-Unity Setup` in the **README** to get the `Robotics` tab on top of your editor. 
![Screenshot 2022-03-13 16:56:44](https://user-images.githubusercontent.com/64059887/158052455-1a6670be-682e-44b1-87dc-b4207ffb3bda.png)
In the screenshot you need to enable 2 things,  one of them is `api compatibility level` set it to **.NET 4.x** and another is to enable unsafe code. 

Okay for the DJI model, currently it is only setup for **PX4**, what you need to control the position is **mavros/local_position/pose**. For my repo, it links up with the **`mavros` topics** and for **multiple agents**, but for you, I think maybe you should start from scratch, since you just need a **simple setup**, you can refer to the **Scene/ViconPointCloudTest** where uav object has S1, S2, S3, and each has a **Ros Mavros Subscriber** script attached to them, and in the uav object it holds the **Ros Connection** script.

In other words : [Unity] ROS Connection (port and IP address) + Your own script (which contains the topic to subscribe or publish) <-> [ROS] TCP endpoint will receive TCP stream, get the messages and stream to ROS topics or from ROS to subscribe and stream to Unity

Hope this helps!

**Comment by ge35tay** [comment]
I made it! Thank u for your quick reply and your suggestions! I use the Tiny drone as the gameobject and now I can get the position of the drone through the ros. And I have a further question, if I want to control the drone in the unity based on ROS, did u have any advise? Did I need to control based on the PX4 simulator?

**Comment by matthewoots** [comment]
That's great! Sadly there's no quadcopter dynamics running on Unity3D so that's why we need an external simulator like Gazebo etc, which is what RotorS or Px4 SITL are using. If you want to do your own, then you'll need work with Unity Physics engine to output forces per rotor and a controller in ROS (it's tough)

**Comment by ge35tay** [comment]
Aha, so basically we can use the PX4, ros and unity at the same time, so when you control the drone in the Gazebo the position of drone in Unity3D will also change synchronously?

**Comment by matthewoots** [comment]
Yup, since in PX4 SITL Gazebo generates sensor data and through PX4 running EKF2, which does sensor fusion, you can get your local position, which is what Unity receives and that is how your Gameobject will move. 

Using Unity is more for running image-base experiments (since it's more realistic) like testing tracking, object detection, depth map algorithms, training NN like what UZH does with https://github.com/uzh-rpg/flightmare

**Comment by ge35tay** [comment]
Yes, i see :) I also tried to build the physical model in Unity3D but it is too hard for me, even with the urdf file of the drone. 

I have installed the PX4 simulator and written a offboard.cpp to control the drone in the gazebo, it looks good now, the drone can run in the expected trajectory now. How can I connect this to Unity3D using mavlink? I have seen your px4-unity-simulator repo but it seems that i can not set up the mavlink with the c# code in the repo. Did you have any reference?

And another question is, should the Gameobject in the Unity3d have the same model with that in Gazebo? Like both is in Drone1.5?

**Comment by ge35tay** [comment]
Hi, sorry to interrupt again. I am trying to connect the Unity3D with Gazebo using ros-tcp-connector. If I understand correctly, in the robotics -> generate ros messages we should generate the msg files in the mavros messages files because the state we need is stored in mavlink topics, am I right?  But when I generate these messages there are even 492 errors. Have you met this problem before? 
![微信图片_20220317164435](https://user-images.githubusercontent.com/65305789/158839754-f9a641f7-1f57-43d3-b6e5-a1c08a118e14.png)

**Comment by matthewoots** [comment]
I haven't been able to replicate the sensor suite so that EKF2 can fuse the data properly, this is why I'm still not able to use Unity3D as the base to connect directly to PX4. If you want to use Unity3d as a simulator you need to use `make px4_sitl none` instead of `gazebo` 

The GameObject doesnt have to be the same, both have almost no communications with each other so there is no need to establish the same model.
